lsmsr/lsmsr.8
.\" Copyright (C) 2008 Advanced Micro Devices, Inc.
'\"! tbl | mmdoc 
'\"macro stdmacro 
.nr X
.TH lsmsr 8 "July 2008" "x86utils"
.SH NAME
lsmsr \- show MSR information for x86 CPUs

.SH SYNOPSIS
.B lsmsr
.RB [ \-hv ]
.RB [ \-a
.RB [ \-c
.IR cpu_nr ]
.RB [ \-f
.IR family ]
.RB [ \-l]
.RB [ \-r
.IR name|address ]
.RB [\-V
.IR verbosity]
.IR [MSR]

.SH DESCRIPTION
.PP
.B lsmsr
is a tool to display information for machine specific registers
of x86 CPUs. Following CPUs are supported:
AMD family 0xf and family 0x10. For Intel and other AMD CPUs just
a some common MSRs is supported.
.SH OPTIONS
.TP
.B -a, --all
Show information for all MSRs known to the tool for that CPU family.
.TP
.B -c cpu_nr, --cpu  cpu_nr
Select CPU (by number) for which MSR information should be displayed
(default: 0).
.TP
.B -f fam, --family fam
Specify CPU family. Normally CPU family is auto-detected. You can
use this option to disable auto-detection, e.g. for debugging
purposes.
.TP
.B -l, --list
Show definition (address, field description) of selected MSR(s).
.TP
.B -r name|address, --register name|address
Specify MSR  (by name or address in hex) for which MSR information
should be displayed.
.TP
.B -V num, --verbosity num
Select verbosity of output format (between 0 and 4, default: 0)
.TP
.B -h, --help
Print help message and exit.
.TP
.B -v, --version
Display version info and exit.
.SH Examples
.IP "" 0
Show all MSRs beginning with 'MTRR' and force CPU family  to 0x10
.IP "" 2
# lsmsr -l -f 0x10 MTRR
 MTRRcap             : 0x000000fe
 MTRRphysBase0       : 0x00000200
 MTRRphysMask0       : 0x00000201
 MTRRphysBase1       : 0x00000202
 MTRRphysMask1       : 0x00000203
 MTRRphysBase2       : 0x00000204
 MTRRphysMask2       : 0x00000205
 MTRRphysBase3       : 0x00000206
 MTRRphysMask3       : 0x00000207
 MTRRphysBase4       : 0x00000208
 MTRRphysMask4       : 0x00000209
 MTRRphysBase5       : 0x0000020a
 MTRRphysMask5       : 0x0000020b
 MTRRphysBase6       : 0x0000020c
 MTRRphysMask6       : 0x0000020d
 MTRRphysBase7       : 0x0000020e
 MTRRphysMask7       : 0x0000020f
 MTRRfix64K_00000    : 0x00000250
 MTRRfix16K_80000    : 0x00000258
 MTRRfix16K_A0000    : 0x00000259
 MTRRfix4K_C0000     : 0x00000268
 MTRRfix4K_C8000     : 0x00000269
 MTRRfix4K_D0000     : 0x0000026a
 MTRRfix4K_D8000     : 0x0000026b
 MTRRfix4K_E0000     : 0x0000026c
 MTRRfix4K_E8000     : 0x0000026d
 MTRRfix4K_F0000     : 0x0000026e
 MTRRfix4K_F8000     : 0x0000026f
 MTRRdefType         : 0x000002ff
.IP ""0
Show 'MTRRcap MSR in verbose one-line mode
.IP "" 2
# lsmsr -r MTRRcap -V 1   
 MTRRcap = 0x0000000000000508 (MtrrCapVCnt=0x8,  MtrrCapFix=0x1,  MtrrCapWc=0x1)
.IP "" 0
Show 'MTRRdefType' in verbose multi-line mode 
.IP "" 2
# lsmsr -r MTRRdefType -V 3 
 MTRRdefType          = 0x0000000000000c00
   MtrrDefMemType=0
   MtrrDefTypeFixEn=0x1
   MtrrDefTypeEn=0x1
.IP "" 0
Show definition of MSR 0x200 in verbose multi-line mode (including reserved fields)
.IP "" 2
# lsmsr -r 0x200 -V 4 -l
 MTRRphysBase0: 0x00000200
   0-7:Type
   8-11:res
   12-39:PhyBase
   40-63:res
.IP "" 0
List all known MSRs for family 0xf
.IP "" 2
# lsmsr  -f 0xf -l -a 
 TSC                 : 0x00000010; time-stamp counter
 APIC_BASE           : 0x0000001b; APIC base address

 ...

 MTRRphysBase0       : 0x00000200; base of variable MTRR (0)
 MTRRphysMask0       : 0x00000201; mask of variable MTRR (0)
 MTRRphysBase1       : 0x00000202; base of variable MTRR (1)
 MTRRphysMask1       : 0x00000203; mask of variable MTRR (1)

 ...

 MTRRfix4K_F0000     : 0x0000026e
 MTRRfix4K_F8000     : 0x0000026f
 PAT                 : 0x00000277; page attribute table
 MTRRdefType         : 0x000002ff


.SH Author
.B lsmsr
and this manual page was written by Andreas Herrmann <andreas.herrman3@amd.com>.
.PP
Permission is granted to copy, distribute and/or modify this 
document under the terms of the GNU General Public License version 2.

.SH REPORTING BUGS
Please send bug reports to <andreas.herrmann3@amd.com>.


ChangeLog
2005-08-07 Dave Jones <davej@codemonkey.org.uk>
Released as v1.14

2005-04-20 Kurt Garloff <garloff@suse.de>

	* cpuid.c, cpuid_UP.S, havecpuid.c: Native support for x86-64.

2005-02-08 Dave Jones <davej@codemonkey.org.uk>
Released as v1.13

2004-12-22 Dave Jones <davej@codemonkey.org.uk>

	* Intel/cachesize.c: Add some EM64T cache descriptors. 
	  Submitted by Kornilios Kourtis <kkourt@cslab.ece.ntua.gr>

2004-10-6 Dominik Brodowski <linux@dominikbrodowski.de>

	* Intel/info.c: report 1 instead of 0 logical CPUs.

2004-8-20 Dave Jones <davej@codemonkey.org.uk>

	* Intel/identify.c: Add 6d6 Pentium-M recognition.

2004-7-11 Dave Jones <davej@codemonkey.org.uk>

	* AMD/identify: Recognise 0xFE0 stepping Athlon64's

2004-6-2 Dave Jones <davej@codemonkey.org.uk>

	* AMD/bugs.c: Fix athlon check.

2004-5-7 Dave Jones <davej@codemonkey.org.uk>

	* AMD/identify.c: Recognise SH8-D0 K8's

2004-5-6 Dave Jones <davej@codemonkey.org.uk>

	* cpuid_UP.S: Add marker for non-executable stack

2004-3-19 Dave Jones <davej@codemonkey.org.uk>

	* AMD/identify.c: Identify a bunch more opterons/athlon64's
	* Intel/info.c: Fix up the number of siblings per package.

2004-2-2 Dave Jones <davej@codemonkey.org.uk>

	* Intel/identify.c: Add P4 Prescott.

2003-12-31 Dave Jones <davej@codemonkey.org.uk>
	* Intel/identify.c: Add P3 Xeon S-Spec numbers.

2003-12-18 Dave Jones <davej@codemonkey.org.uk>
	* Intel/identify.c: Add P4 S-Spec numbers.
	Add recognition note for P4 extreme edition.

2003-11-4 Dave Jones <davej@codemonkey.org.uk>

	* AMD/identify.c: Add various random CPU facts.
	* IDT/*.c: Various cleanup of powersaver/longhaul decoding.
	* identify.c: Code cleanup.

2003-9-24 Dave Jones <davej@codemonkey.org.uk>

	* Intel/identify.c: Fix P4 errata url.

2003-9-12 Dave Jones <davej@codemonkey.org.uk>

	* AMD/identify.c: Opteron B0

2003-8-17 Dave Jones <davej@codemonkey.org.uk>

	* Intel/cachesize.c: Output sorted descriptors.

Released as 1.12b

2003-7-15 Dave Jones <davej@codemonkey.org.uk>

	* IDT/longhaul-v2.c: Use bitfields.
	* binary.c: Add binary() which decodes n bits.

2003-6-13 Jeremy Fitzhardinge <jeremy@goop.org>

	* features.c: identify Intel extended flags.
	* Intel/info.c: Banias also has model name.
	* get_model_name.c: Eat leading spaces.

2003-6-12 Dave Jones <davej@codemonkey.org.uk>

	* Intel/cachesize.c: print out unknown descriptors again.
	  Don't iterate around loop once we've found the descriptor.

2003-6-12 Jeremy Fitzhardinge <jeremy@goop.org>

	* Intel/MSR-PM.c: Decode PERF_STATUS.

2003-6-8 Dave Jones <davej@codemonkey.org.uk>

	* Intel/Intel.h, Intel/cachesize.c, Intel/identify.c, Intel/info.c:
	  Make the cachesize output look like it did in previous releases.
	  This is quite ugly, but is still a step forward to how I want it
	  to look one day.
	* bench/MHz.c: Rounding up to nearest 50MHz if MHz is within a 15MHz
	  discrepancy of the nearest 50. Crap, needs improving.
	* x86info.c: Make GHz output look like 2.8GHz instead of 2.800GHz
	  disable --bench for this release.
	* rdmsr.c: dump val as 32 bit value if passed 32 bit size.

2003-6-8 Bryan O'Sullivan <bos@serpentine.com>

	* AMD/identify.c: Opteron is socket 940 not 754
	* x86info.h, connector.c: Add Socket 940

2003-6-9 Jeremy Fitzhardinge <jeremy@goop.org>

	* Intel/info.c: Pentium M MSR support.
	* Intel/MSR-PM.c: Pentium M MSR support.
	* Makefile: Pentium M MSR support.
	* Intel/Intel.h: Pentium M MSR support.
	* Intel/identify.c: s/Centrino/Banias/

2003-5-3 Dave Jones <davej@codemonkey.org.uk>

	* Intel/identify.c: Add S codes for Pentium II.

2003-5-2 Dave Jones <davej@codemonkey.org.uk>

	* identify.c: Display MHz as GHz if >1000. Always do MHz calculation,
	  even if no --MHz is specified
	* Mhz.c: Store results in cpu struct. Use integer instead of fp.
	* Intel/identify.c: We can further decode sSpec if we use MHz.

2003-4-24 Dave Jones <davej@codemonkey.org.uk>

	* Intel/identify.c: Add decoded ppro sSpec numbers as comments.

2003-4-23 Dave Jones <davej@codemonkey.org.uk>

	* Intel/identify.c: Add extra space between Pentium Pro and stepping.

2003-4-22 Dave Jones <davej@codemonkey.org.uk>

	* AMD/identify.c: Add Opteron stepping from public docs.

2003-4-11 Dave Jones <davej@codemonkey.org.uk>

	* SiS/identify.c: Add recognition of SiS 55x
	* rdmsr.c: when dumping long long zero pad to 64 bit.
	* Intel/MSR-P4.c: Zero pad to 64 bit.
	* Intel/bluesmoke.c: Zero pad.
	* AMD/bluesmoke.c: Zero pad.
	* AMD/bugs.c: Zero pad.
	* IDT/identifty.c: More interesting comments.

2003-4-2 Dave Jones <davej@codemonkey.org.uk>

	* IDT/powersaver.c: Fix decoding of RevisionID

2003-3-28 Dave Jones <davej@codemonkey.org.uk>

	* Intel/cachesize.c: Add Centrino cache descriptors.
	* Intel/identify.c: Add P4-M (Centrino)

2003-3-19 Dave Jones <davej@codemonkey.org.uk>

	* Intel/cachesize.c: Add 0x3b descriptor.

2003-3-14 Dave Jones <davej@codemonkey.org.uk>

	* identify.c: decode 0xc0000000 range cpuids if available.
	* bench/bench.h: improve timing routine a little. still needs work.
	* bench/benchmarks.c: remove lea (eax), 1(eax) test for now.

2003-3-13 Dave Jones <davej@codemonkey.org.uk>

	* Intel/identify.c: Recognise P4 based Celerons.

2003-3-11 Dave Jones <davej@codemonkey.org.uk>

	* AMD/powernow.c: Decode extra bits present in x86-64
	* AMD/identify.c: Add some extra x86-64 steppings.

2003-3-7 Dave Jones <davej@codemonkey.org.uk>

	* Intel/identify.c: Recognise D-1 stepping P4s.

2003-2-11 Dave Jones <davej@codemonkey.org.uk>

	* Intel/microcode.c: Add leading 0x to version output

2003-2-10 Dave Jones <davej@codemonkey.org.uk>

	* IDT/identify.c:
	* Intel/identify.c: Add various comments that may later
	  be used as further info to print out.

2003-2-7 Dave Jones <davej@codemonkey.org.uk>

	* bench/benchmarks.c: Add bswap.
	* Makefile: Use -m32 when compiling on x86-64
	* x86info.c: Bump copyright date.

2003-1-27 Dave Jones <davej@codemonkey.org.uk>

	* scripts/testnodes: Only check for # of CPUs present.

lsmsr/lsmsr.c
/*
 * Author: Andreas Herrmann <andreas.herrmann3@amd.com>
 *
 * Copyright (C) 2008 Advanced Micro Devices, Inc.
 *
 * Licensed under the terms of the GNU GENERAL PUBLIC LICENSE version 2.
 * See file COPYING for details.
 */

#define _LARGEFILE64_SOURCE
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>

#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif
#include <getopt.h>

#include <msr.h>
#include <x86info.h>
#include "AMD_k8.h"
#include "AMD_fam10h.h"
#include "AMD_fam11h.h"
#include "AMD_fam12h.h"
#include "AMD_fam14h.h"
#include "AMD_fam15h.h"
#include "generic_msr.h"

/* Todos:
 * - add (list and eventually write) support for write-only MSR
 * - add decoding support for bit fields
 * - proper check for MSR support
 * - add support for generic MSRs for non-Intel/AMD CPUs
 */

#define LSMSR_VERSION "0.816"

struct {
	int fd;
	int show_all;
	int list;
	int verbosity;
	int cpu;
	enum vendor vendor;
	int family;
	int model;
	struct reg_spec *msr_table;
	const char *prog;
	const char *msr_name;
	const char *msr_search;
	uint32_t reg;
} g = {
	.reg = -1,
};


/* getopt_long stuff */
static struct option lopts[] = {
	{"all", no_argument, 0, 'a'},		/* show all MSRs */
	{"cpu", required_argument, 0, 'c'},	/* CPU number */
	{"family", required_argument, 0, 'f'},	/* CPU family */
	{"list", no_argument, 0, 'l'},		/* list known MSRs */
	{"register", required_argument, 0, 'r'},  /* register name or address */
	{"help", no_argument, 0, 'h'},		/* help */
	{"model", required_argument, 0, 'm'},	/* CPU model */
	{"version", no_argument, 0, 'v'},	/* version info */
	{"verbosity", required_argument, 0, 'V'}, /* verbosity */
	{0, 0, 0, 0}
};

#define OPTSTRING "ac:C:f:hlm:r:vV:"

#define _USAGE "[OPTION] <MSR>\n"

#define _HELP \
"  -a, --all                  show info for all known MSRs\n"\
"  -c, --cpu <CPU>            specify CPU for which MSRs are shown (default 0)\n"\
"  -f, --family <cpu family>  set cpu family to be used\n"\
"  -l, --list                 list info about MSR(s), its addresses and fields\n"\
"  -r, --register <name|addr> select register by name or address\n"\
"  -h, --help                 show help\n"\
"  -m, --model                set cpu model to be used\n"\
"  -v, --version              show version info\n"\
"  -V, --verbosity <n>        set verbosity\n"\
"                             V=1 show all fields of the register\n"\
"                             V=2 dito and show reserved fields\n"\
"                             V=3 like V=1 but each field on separate line\n"\
"                             V=4 dito and show reserved fields\n"\
"\n"\
" Display model specific registers on AMD64 processors.\n"\
" If <MSR> is given information for all MSRs for which <MSR> matches the\n"\
" beginning of its name are displayed. Precedence of MSR selection is:\n"\
" \" -r <name|addr>\" overrules \"<MSR>\" overrules \"-a\".\n"\
"\n"

static void usage(void)
{
	fprintf(stderr, "Usage: %s "_USAGE, g.prog);
	fprintf(stderr, "Try '%s --help' for more information\n", g.prog);
	exit(1);
}

static void help(void)
{
	fprintf(stdout, "Usage: %s "_USAGE, g.prog);
	fprintf(stdout, "Help:\n"_HELP);
}

static void version(void)
{
	fprintf(stdout, "%s version %s\n", g.prog, LSMSR_VERSION);
}

static int get_msr_val(unsigned int msr, unsigned long long *val)
{
	off64_t off;
	int err;

	*val = 0;
	off = lseek64(g.fd, (off64_t) msr, SEEK_SET);
	if (off == (off_t) -1) {
		perror("invalid MSR");
		return 1;
	}

	off = read(g.fd, val, 8);
	err = errno;
	if (off != 8) {
		fflush(stdout);
		fprintf(stderr,
			"could not read MSR 0x%8.8x (%s): %s\n",
			msr, get_reg_name(msr, g.msr_table), strerror(err));
		return 0;
	}

	return 0;
}

static int open_dev(int cpu)
{
	char s[20];

	snprintf(s, sizeof(s), "/dev/cpu/%d/msr", cpu);
	g.fd = open(s, O_RDONLY);
	if (g.fd < 0)
		fprintf(stderr, "could not open device %s: %s\n", s,
			strerror(errno));

	return g.fd;
}

#define PRINT_MSR(reg, val) \
do { \
	print_reg(reg, val, g.list, g.show_all, g.verbosity); \
} while(0)

static int _show_msr(struct reg_spec *reg)
{
	unsigned long long val = 0;
	if (!g.list)
		if (get_msr_val(reg->address, &val))
			return 1;
	PRINT_MSR(reg, val);
	return 0;
}

static int show_matching(const char *name)
{
	int i, t;
	int n = MSR_MAX_LEN;
	int ret = 0;

	t = strlen(name);
	if (t < n)
		n = t;

	g.show_all = 1;
	for (i = 0; g.msr_table[i].name; i++) {
		t = strlen(g.msr_table[i].name);
		if (n < t)
			t = n;
		if (t < n)
			continue;
		if (strncmp(name, g.msr_table[i].name, t) == 0)
			if (_show_msr(&(g.msr_table[i]))) {
				ret = 1;
				break;
			}
	}
	g.show_all = 0;
	return ret;
}

struct vendor_string {
	enum vendor id;
	const char *name;
} vendor_names[] = {
	{VENDOR_UNKNOWN, "(unknown)"},
	{VENDOR_AMD, "AMD"},
	{VENDOR_CENTAUR, "Centaur"},
	{VENDOR_CYRIX, "Cyrix"},
	{VENDOR_INTEL, "Intel"},
	{VENDOR_NATSEMI, "Natsemi"},
	{VENDOR_RISE, "Rise"},
	{VENDOR_TRANSMETA, "Transmeta"},
	{VENDOR_SIS, "SIS"},
};
get_name(vendor, enum vendor, vendor_names);

static void set_vendor(void)
{
	unsigned int b;
	cpuid(g.cpu, 0, NULL, &b, NULL, NULL);

	switch (b) {
	case 0x68747541:
		g.vendor = VENDOR_AMD;
		break;
	case 0x756e6547:
		g.vendor = VENDOR_INTEL;
		break;
	case 0x69727943:
		g.vendor = VENDOR_CYRIX;
		break;
	case 0x746e6543:
		g.vendor = VENDOR_CENTAUR;
		break;
	case 0x646f6547:
		g.vendor = VENDOR_NATSEMI;
		break;
	case 0x52697365:
	case 0x65736952:
		g.vendor = VENDOR_RISE;
		break;
	case 0x20536953:
		g.vendor = VENDOR_SIS;
		break;
	default:
		g.vendor = VENDOR_UNKNOWN;
		break;
	}
}

static void set_family_model(void)
{
	unsigned int a;
	cpuid(g.cpu, 0, &a, NULL, NULL, NULL);
	if (!g.family) {
		cpuid(g.cpu, 1, &a, NULL, NULL, NULL);
		g.family = (a >> 8) & 0xf;
		if (g.family == 0xf)
			g.family += (a >> 20) & 0xff;
	}

	if (!g.model) {
		cpuid(g.cpu, 1, &a, NULL, NULL, NULL);
		g.model = (a >> 4) & 0xf;
		if (g.model == 0xf)
			g.model |= (a >> 12) & 0xf0;
	}
}

static int set_msr_table(void)
{
	int supported = 0;

	set_vendor();
	set_family_model();

	if (g.vendor == VENDOR_AMD) {
		supported = 1;
		switch (g.family) {
		case 0x0f:
			g.msr_table = AMD_k8_spec;
			break;
		case 0x10:
			g.msr_table = AMD_fam10h_spec;
			break;
		case 0x11:
			g.msr_table = AMD_fam11h_spec;
			break;
		case 0x12:
			g.msr_table = AMD_fam12h_spec;
			break;
		case 0x14:
			g.msr_table = AMD_fam14h_spec;
			break;
		case 0x15:
			g.msr_table = AMD_fam15h_spec;
			break;
		default:
			g.msr_table = generic_msr_spec;
		}
	}

	if (g.vendor == VENDOR_INTEL) {
		supported = 1;
		g.msr_table = generic_msr_spec;
	}

	if (!supported) {
		fprintf(stdout, "CPU not (yet) supported "
			"(vendor=\"%s\", family=%d, model=%d)\n",
			get_vendor_name(g.vendor), g.family, g.model);
		return 1;
	}

	return 0;
}

struct reg_spec unknown_msr = {0, "unknown", "(at your own risk)", NULL, NULL};

#define OPT_MAX 32
int main(int argc, char *argv[])
{
	char c;
	int i, li, ret;
	struct reg_spec *reg;

	ret = 1;
	if((g.prog = rindex(argv[0], '/')))
		++g.prog;
	else
		g.prog = argv[0];

	if (!HaveCPUID()) {
		fprintf(stderr, "warning: no cpuid instruction available\n");
		fprintf(stdout, "no MSR information available for this CPU\n");
		return 0;
	}

	while((c = getopt_long(argc, argv, OPTSTRING, lopts, &li)) != -1) {
		switch (c) {
		case 'a':
			g.show_all = 1;
			break;
		case 'h':
			help();
			return 0;
		case 'v':
			version();
			return 0;
		case 'V':
			g.verbosity = strtol(optarg, NULL, 10);
			break;
		case 'c':
			g.cpu = strtol(optarg, NULL, 0);
			break;
		case 'm':
			g.model = strtol(optarg, NULL, 0);
			break;
		case 'f':
			g.family = strtol(optarg, NULL, 0);
			break;
		case 'l':
			g.list = 1;
			break;
		case 'r':
			if (isalpha(*optarg))
				g.msr_name = optarg;
			else
				g.reg = (unsigned int) strtol(optarg, NULL, 16);
			break;
		default:
			usage();
		}
	}

	if (argc < 2) {
		usage();
		goto out;
	}

	if ((optind > -1) && argv[optind]) {
		if ((argc - optind) != 1)
			usage();
		if (strlen(argv[optind]) >= OPT_MAX) {
			fprintf(stderr,"error: invalid command line\n");
			goto out;
		}
		g.msr_search = argv[optind];
	}

	if (set_msr_table())
		goto out;

	if (g.msr_name) {
		g.reg = get_reg_addr(g.msr_name, g.msr_table);
		if (g.reg == (uint32_t) -1) {
			fflush(stdout);
			fprintf(stderr, "error: unknown register \"%s\"\n",
				g.msr_name);
			goto out;
		}
	}

	if (!g.list)
		if(open_dev(g.cpu) < 0)
			goto out;

	if (g.reg != (uint32_t) -1) {
		reg = get_reg_spec(g.reg, g.msr_table);
		if (!reg) {
			fflush(stdout);
			fprintf(stderr, "warning: unknown MSR %x\n", g.reg);
			g.verbosity = 0;
			unknown_msr.address = g.reg;
			reg = &unknown_msr;
		}
		if (_show_msr(reg))
			goto out;
	} else 	if (g.msr_search) {
		if (show_matching(g.msr_search))
			goto out;
	} else if (g.show_all) {
		for (i = 0; g.msr_table[i].name; i++)
			if (_show_msr(&(g.msr_table[i])))
				goto out;
	}
	ret = 0;

 out:
	if (g.fd >=0)
		close(g.fd);
	return ret;
}
